src/si/helpers.stanza

```diff
@@ -88,6 +89,143 @@ public defn topo-net (pair:KeyValue<JITXObject,JITXObject|KeyValue>):
+
+  Given a sequence of key-value pairs of "bus segments", this function
+  will connect them, generate topology segments, and constrain the trace
+  length as necessary.
+
+  A "bus segment" is either:
+  * A tuple of single-pin or array ports
+  * A single-pin or array port
+  * An Instantiable
+  * A "ShortTrace"
+
+  The left- and right-most components of the chain must be concrete --
+  either a pin or a tuple of pins. If given an Instantiable
+  for one segment of the chain, this function will create as many concrete
+  instances as necessary.
+
+  We do not currently accept concrete instances, just pins or Instantiables.
+
+  We assume that any compatible Instantiable has pins `p[1]` and `p[2]`.
+
+  We assume that all concrete segments are the same size and shape:
+  we can only compute `topo-bus([a, b, c], [d, e, f])` when port `a`
+  matches port `d`, and so on.
+
+  Returns a tuple of all newly-instantiated objects.
+
+  @snippet
+
+  ```stanza
+  port mii : rmii-bus
+  require sw-bus : rmii-bus from mcu
+  val series-R-type = create-resistor(resistance = 51.0, precision = (1 %), case = ["0603"])
+  val rx-bus = [sw-bus.rxd, sw-bus.crs-dv, sw-bus.ref-clk]
+  val rx-out = [mii.rxd, mii.crs-dv, mii.ref-clk]
+  topo-bus(ShortTrace(rx-bus => series-R-type) => rx-out)
+  ```
+
+  @snip-note 3 `create-resistor` returns an `Instantiable`
+  @snip-note 4 `sw-bus.rxd` and `mii.rxd` are both arrays of two pins. `topo-bus` will return a tuple of four new resistors that connect the two buses.
+  @snip-note 6 All objects in `rx-bus` will be connected to their respective resistor with a short-trace.
+<DOC>
+public defn topo-bus (pair:KeyValue) -> Tuple<JITXObject> :
+  ensure-valid-tree!(pair)
+  val [l, r] = unpack(pair)
+  val acc = Vector<JITXObject>()
+  make-topo-bus(l, r, acc, short-trace? = pair is ShortTrace)
+  to-tuple $ acc
+
+defn ensure-valid-tree! (kv:KeyValue) : false ; [TODO] : check precondition
+
+; Traverse left, then connect to the right
+defn make-topo-bus (l:KeyValue|JITXObject|Tuple, r:KeyValue|JITXObject|Tuple|Instantiable, acc:Vector, -- short-trace?:True|False = false) -> Tuple<JITXObject> :
+  val lhs = flatten(l, acc)
+  connect(lhs, r, acc, short-trace?)
+
+defn connect (lhs:Tuple, r:KeyValue|JITXObject|Tuple|Instantiable, acc:Vector, short-trace?) -> Tuple :
+  match(r) :
+    (x:KeyValue) :
+      val [l, r] = unpack(x)
+      val lhs* = connect(lhs, l, short-trace?, acc)
+      connect(lhs*, r, acc, x is ShortTrace)
+    (x:JITXObject) :
+      connect-topo-bus(lhs, flatten-obj(x), short-trace?)
+    (x:Tuple) :
+      val r* = to-tuple $ seq-cat(flatten-obj, x)
+      connect-topo-bus(lhs, r*, short-trace?)
+    (x:Instantiable) :
+      connect-topo-bus(lhs, x, acc, short-trace?)
+
+defn connect-topo-bus (lhs:Tuple, rhs:Instantiable, acc, short-trace?:True|False) -> Tuple :
+  inside pcb-module :
+    to-tuple $
+      for l in lhs seq :
+        inst i : rhs 
+        add(acc, i)
+        topo-net(l, i.p[1])
```

Comment:
Let's use `get-element-ports` here so that we can use both `p[1]/p[2]` and `a/c` variant two pins.

Reply:
If we do that, we may want to have some mechanism to invert - perhaps we leave that the `Network` code and the user can just instantiate a network with a single diode and an invert:

```
  val flip-diode = create-circuit( (~ D))
  topo-bus(rx-bus => flip-diode => out-bus)
```

Or something similar. 

I guess the thing that I want to avoid is if a user puts a Diode into this `topo-bus` chain - I don't want them to get some `Can't find port p[1]` error  that they won't be able to discern.